### PR TITLE
Use Flatbuffer Field type instead of MaterializedField

### PR DESCRIPTION
**Pull request by StevenMPhillips**
Remove MaterializedField, MajorType, RepeatedTypes

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/Types.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.arrow.vector;
+
+import com.google.common.base.Preconditions;
+import com.google.flatbuffers.FlatBufferBuilder;
+import org.apache.arrow.flatbuf.Field;
+import org.apache.arrow.flatbuf.Int;
+import org.apache.arrow.flatbuf.Type;
+
+public class Types {
+  public static void main(String[] args) {
```
Comment (kiril-me) [inline java/vector/src/main/java/org/apache/arrow/vector/Types.java]: Do we need this class?

**Comment by julienledem** [comment]
@StevenMPhillips can you prefix the PR by the JIRA? (ARROW-XXX: ...)

File: format/Message.fbs [hunk 2]
```diff
@@ -1,10 +1,13 @@
-namespace apache.arrow.flatbuf;
+namespace org.apache.arrow.flatbuf;
 
 /// ----------------------------------------------------------------------
 /// Logical types and their metadata (if any)
 ///
 /// These are stored in the flatbuffer in the Type union below
 
+table Null {
+}
```
Comment (julienledem) [inline format/Message.fbs]: What does a Null type vector layout look like?

File: java/vector/src/main/codegen/templates/BaseReader.java [hunk 3]
```diff
@@ -30,8 +30,8 @@
 
 @SuppressWarnings("unused")
 public interface BaseReader extends Positionable{
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/BaseReader.java]: could you add javadoc on interfaces methods?

File: java/vector/src/main/codegen/templates/AbstractFieldReader.java [hunk 4]
```diff
@@ -41,7 +41,13 @@ public boolean isSet() {
     return true;
   }
 
-  <#list ["Object", "BigDecimal", "Integer", "Long", "Boolean", 
+  @Override
+  public Field getField() {
+    fail("getField");
+    return null;
+  }
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/AbstractFieldReader.java]: do we need this method? In what case is it not overridden?

File: java/vector/src/main/codegen/templates/BaseWriter.java [hunk 5]
```diff
@@ -31,12 +31,11 @@
  */
 @SuppressWarnings("unused")
   public interface BaseWriter extends AutoCloseable, Positionable {
-  FieldWriter getParent();
   int getValueCapacity();
 
   public interface MapWriter extends BaseWriter {
 
-    MaterializedField getField();
+    Field getField();
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/BaseWriter.java]: javadoc on interfaces?

File: java/vector/src/main/codegen/templates/HolderReaderImpl.java [hunk 6]
```diff
@@ -176,10 +134,6 @@ public void read(Nullable${name}Holder h) {
       Period p = new Period();
       return p.plusDays(holder.days).plusMillis(holder.milliseconds);
 
-<#elseif minor.class == "Decimal9" ||
-         minor.class == "Decimal18" >
-      BigInteger value = BigInteger.valueOf(holder.value);
-      return new BigDecimal(value, holder.scale);
 
 <#elseif minor.class == "Decimal28Dense" ||
          minor.class == "Decimal38Dense">
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/HolderReaderImpl.java]: is this used?

File: java/vector/src/main/codegen/templates/UnionListWriter.java [hunk 7]
```diff
@@ -64,10 +63,14 @@ public void clear() {
   }
 
   @Override
-  public MaterializedField getField() {
+  public Field getField() {
     return null;
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/UnionListWriter.java]: throw ? 
what does null mean in this context?

File: java/vector/src/main/codegen/templates/UnionListWriter.java [hunk 8]
```diff
@@ -91,7 +100,7 @@ public void close() throws Exception {
 
   @Override
   public ${name}Writer <#if uncappedName == "int">integer<#else>${uncappedName}</#if>(String name) {
-    assert inMap;
+//    assert inMap;
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/UnionListWriter.java]: what was this for?

File: java/vector/src/main/codegen/templates/UnionVector.java [hunk 9]
```diff
@@ -47,69 +56,51 @@
...
 
   private int singleType = 0;
   private ValueVector singleVector;
-  private MajorType majorType;
 
   private final CallBack callBack;
 
-  public UnionVector(MaterializedField field, BufferAllocator allocator, CallBack callBack) {
-    this.field = field.clone();
+  public UnionVector(String name, BufferAllocator allocator, CallBack callBack) {
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/UnionVector.java]: can you add javadoc?
In particular. What's the contract of the callback?

File: java/vector/src/main/codegen/templates/UnionVector.java [hunk 10]
```diff
@@ -47,69 +56,51 @@
...
   public MapVector getMap() {
     if (mapVector == null) {
       int vectorCount = internalMap.size();
-      mapVector = internalMap.addOrGet("map", MAP_TYPE, MapVector.class);
-      addSubType(MinorType.MAP);
+      mapVector = internalMap.addOrGet("map", MinorType.MAP, MapVector.class);
       if (internalMap.size() > vectorCount) {
         mapVector.allocateNew();
+        if (callBack != null) {
+          callBack.doWork();
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/UnionVector.java]: CallBack looks like Runnable.

File: java/vector/src/main/codegen/templates/UnionVector.java [hunk 11]
```diff
@@ -196,33 +190,40 @@ public void close() {
...
     internalMap.clear();
   }
 
   @Override
-  public MaterializedField getField() {
-    return field;
+  public Field getField() {
+    List<org.apache.arrow.vector.types.pojo.Field> childFields = new ArrayList<>();
+    for (ValueVector v : internalMap.getChildren()) {
+      if (v.getField().getName().equals("types")) {
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/UnionVector.java]: can you add a comment regarding the "types" vector and why we skip it.

File: java/vector/src/main/codegen/templates/VariableLengthVectors.java [hunk 12]
```diff
@@ -68,16 +66,42 @@
...
+    super(name, allocator);
     this.oAccessor = offsetVector.getAccessor();
     this.accessor = new Accessor();
     this.mutator = new Mutator();
   }
+  </#if>
+
+  @Override
+  public Field getField() {
+        throw new UnsupportedOperationException("internal vector");
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/VariableLengthVectors.java]: fix indentation.

File: java/vector/src/main/codegen/templates/VariableLengthVectors.java [hunk 13]
```diff
@@ -125,27 +149,6 @@ public int getVarByteLength(){
...
-//     the bits vector is the first child (the order in which the children are added in getMetadataBuilder is significant)
-//    final SerializedField offsetField = metadata.getChild(0);
-//    offsetVector.load(offsetField, buffer);
-//
-//    final int capacity = buffer.capacity();
-//    final int offsetsLength = offsetField.getBufferLength();
-//    data = buffer.slice(offsetsLength, capacity - offsetsLength);
-//    data.retain();
-//  }
-
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/VariableLengthVectors.java]: remove?

File: java/vector/src/main/java/org/apache/arrow/vector/Types.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    int field = Field.createField(b, nameOffset, true, Type.Int, typeOffset, childrenOffset);
+    b.finish(field);
+
+    Field f = Field.getRootAsField(b.dataBuffer());
+    assert "name".equals(f.name());
+    Int integer = (Int) f.type(new Int());
+    Preconditions.checkState(f.typeType() == Type.Int);
+    Preconditions.checkState(integer.bitWidth() == 32);
+    Preconditions.checkState(f.childrenLength() == 0);
+  }
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/Types.java]: should this be a unit test?

File: java/vector/src/main/java/org/apache/arrow/vector/ValueVector.java [hunk 15]
```diff
@@ -94,7 +94,9 @@
   /**
    * Get information about how this field is materialized.
    */
-  MaterializedField getField();
+  Field getField();
+
+  MinorType getMinorType();
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/ValueVector.java]: javadoc

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java [hunk 16]
```diff
@@ -17,26 +17,28 @@
...
-import org.apache.arrow.vector.types.MaterializedField;
-import org.apache.arrow.vector.types.Types;
 import org.apache.arrow.vector.types.Types.MinorType;
+import org.apache.arrow.vector.types.pojo.ArrowType.Null;
+import org.apache.arrow.vector.types.pojo.Field;
 import org.apache.arrow.vector.util.TransferPair;
 
 import com.google.common.collect.Iterators;
 
 public class ZeroVector implements ValueVector {
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: can you add javadoc on this one?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/AbstractContainerVector.java [hunk 17]
```diff
@@ -43,12 +41,12 @@
...
   static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(AbstractContainerVector.class);
 
-  protected MaterializedField field;
+  protected final String name;
   protected final BufferAllocator allocator;
   protected final CallBack callBack;
 
-  protected AbstractContainerVector(MaterializedField field, BufferAllocator allocator, CallBack callBack) {
-    this.field = Preconditions.checkNotNull(field);
+  protected AbstractContainerVector(String name, BufferAllocator allocator, CallBack callBack) {
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/AbstractContainerVector.java]: can you add javadoc on when the callback is called?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/AbstractMapVector.java [hunk 18]
```diff
@@ -109,16 +100,15 @@ public boolean allocateNewSafe() {
    *   </li>
    * </ul>
    *
-   * @param name name of the field
-   * @param type type of the field
+   * @param name the field
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/AbstractMapVector.java]: type was renamed to minorType

File: java/vector/src/main/java/org/apache/arrow/vector/types/Types.java [hunk 19]
```diff
@@ -17,150 +17,508 @@
...
+      public ValueVector getNewVector(String name, BufferAllocator allocator, CallBack callBack, int... precisionScale) {
+        return new NullableVarBinaryVector(name, allocator);
+      }
+
+      @Override
+      public FieldWriter getNewFieldWriter(ValueVector vector) {
+        return new VarBinaryWriterImpl((NullableVarBinaryVector) vector);
+      }
+    },   //  variable length binary
+    DECIMAL(null) {
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/types/Types.java]: not supported?

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+    }
+    listVector.clear();
+  }
+
+  @Test
+  public void simpleUnion() {
+    UnionVector vector = new UnionVector("union", allocator, null);
+    UnionWriter unionWriter = new UnionWriter(vector);
+    unionWriter.allocate();
+    for (int i = 0; i < 10000; i++) {
```
Comment (julienledem) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: do we need 10000 entries? or fewer would suffice?

**Comment by julienledem** [comment]
Overall this looks good. I have added a few comments.

**Comment by julienledem** [comment]
@StevenMPhillips can you create a jira for this?

**Comment by julienledem** [comment]
@StevenMPhillips Since this has been merged, can you close it?
Also could you reply to my comments and post a follow up PR?

**Comment by julienledem** [comment]
merged in https://github.com/apache/arrow/commit/e7e399db5fc6913e67426514279f81766a0778d2

**Pull request closed without merging**